macros/src/query.rs

```diff
@@ -193,14 +208,276 @@ pub fn derive(input: DeriveInput) -> Result<TokenStream2> {
+
+    for variant in &data.variants {
+        let (fields, queries) = match variant.fields {
+            syn::Fields::Named(ref fields) => fields
+                .named
+                .iter()
+                .map(|f| {
+                    (
+                        syn::Member::Named(f.ident.clone().unwrap()),
+                        query_ty(&lifetime, &f.ty),
+                    )
+                })
+                .unzip(),
+            syn::Fields::Unnamed(ref fields) => fields
+                .unnamed
+                .iter()
+                .enumerate()
+                .map(|(i, f)| {
+                    (
+                        syn::Member::Unnamed(syn::Index {
+                            index: i as u32,
+                            span: Span::mixed_site(),
+                        }),
+                        query_ty(&lifetime, &f.ty),
+                    )
+                })
+                .unzip(),
+            syn::Fields::Unit => (Vec::new(), Vec::new()),
+        };
+
+        let ident = variant.ident.clone();
+
+        if ident == "__HecsDanglingFetch__" {
+            return Err(Error::new_spanned(
+                ident,
+                "derive(Query) reserves this identifier for internal use",
+            ));
+        }
+
+        let named_fields = fields
+            .iter()
+            .map(|x| match x {
+                syn::Member::Named(ref ident) => ident.clone(),
+                syn::Member::Unnamed(ref index) => {
+                    Ident::new(&format!("field_{}", index.index), Span::mixed_site())
+                }
+            })
+            .collect::<Vec<_>>();
+
+        let fetches = queries
+            .iter()
+            .map(|ty| quote! { <#ty as ::hecs::Query>::Fetch })
+            .collect::<Vec<_>>();
+
+        if dangling_constructor.is_none() && fields.is_empty() {
+            dangling_constructor = Some(quote! {
+                Self::#ident {}
+            });
+        }
+
+        fetch_variants.extend(quote! {
+            #ident {
+                #(
+                    #named_fields: #fetches,
+                )*
+            },
+        });
+
+        state_variants.extend(quote! {
+            #ident {
+                #(
+                    #named_fields: <#fetches as ::hecs::Fetch>::State,
+                )*
+            },
+        });
+
+        query_get_variants.extend(match variant.fields {
+            syn::Fields::Named(_) | syn::Fields::Unnamed(_) => quote! {
+                Self::Fetch::#ident { #(#named_fields),* } => {
+                    #(
+                        let #named_fields: <#queries as ::hecs::Query>::Item<'q> = <#queries as ::hecs::Query>::get(#named_fields, n);
+                    )*
+                    Self::Item::#ident { #( #fields: #named_fields,)* }
+                },
+            },
+            syn::Fields::Unit => quote! {
+                Self::Fetch::#ident {} => Self::Item::#ident,
+            },
```

Comment:
Do these need to be separate branches at all? If `named_fields` is empty, the first branch would generate `Self::Fetch::#ident {} => Self::Item::#ident {}`, which should be perfectly legal.

Reply:
Ah, I didn't know you could construct a unit variant with empty braces.
[27b9eea](https://github.com/Ralith/hecs/pull/384/commits/27b9eea26af3077d6139c04fe60562d8fe1c8902)